File: backend/services/stigmer-server/pkg/runnerfailure/runnerfailure_test.go
package runnerfailure

import (
	"errors"
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.temporal.io/sdk/temporal"
)

// canceledWithMessage mirrors how the SDK's failure converter materializes a
// TS CancelledFailure in Go: NewCanceledErrorWithOptions with the failure
// proto's message (the bare NewCanceledError constructor carries details only).
func canceledWithMessage(msg string) error {
	return temporal.NewCanceledErrorWithOptions(temporal.CanceledErrorOptions{Message: msg})
}

// The runner's own shutdown classification (execute-cursor / execute-deep-agent
// throw this exact message) must be recognized however it crosses the boundary.
func TestIsWorkerShutdown_RunnerClassifiedCanceledFailure(t *testing.T) {
	err := canceledWithMessage("Activity cancelled (worker shutdown, not user pause)")
	assert.True(t, IsWorkerShutdown(err))
}

// The Temporal TS worker's drain text — the exact shape captured live in the
// 2026-08-08 incident — crosses as a non-canceled failure and must map.
func TestIsWorkerShutdown_DrainText(t *testing.T) {
	err := temporal.NewApplicationError(
		"Worker is shutting down and this activity did not complete in time", "")
	assert.True(t, IsWorkerShutdown(err))
}

// A wrapped canceled failure (e.g. inside an activity error chain) is still
// recognized — the recognizer walks the Unwrap chain.
func TestIsWorkerShutdown_WrappedCanceledFailure(t *testing.T) {
	err := fmt.Errorf("activity error: %w", canceledWithMessage("worker shutdown"))
	assert.True(t, IsWorkerShutdown(err))
}

// A user pause is a canceled failure too — it must NOT classify as shutdown.
func TestIsWorkerShutdown_PauseIsNotShutdown(t *testing.T) {
	err := canceledWithMessage("Activity paused by orchestrator")
	assert.False(t, IsWorkerShutdown(err))
}

// The loose markers apply ONLY to canceled failures (runner-authored
// messages). An application error that merely echoes "shutting down" —
// plausible in agent/tool output — must not be mistaken for a worker drain.
func TestIsWorkerShutdown_LooseMarkerDoesNotFireOnApplicationErrors(t *testing.T) {
	err := temporal.NewApplicationError(
		"tool failed: the database is shutting down for maintenance", "")
	assert.False(t, IsWorkerShutdown(err))
}

func TestIsWorkerShutdown_UnrelatedFailures(t *testing.T) {
	assert.False(t, IsWorkerShutdown(errors.New("connection refused")))
	assert.False(t, IsWorkerShutdown(temporal.NewApplicationError("model call failed", "")))
	assert.False(t, IsWorkerShutdown(nil))
}

File: backend/services/stigmer-server/src/temporal/__tests__/runner-failure.test.ts
/**
 * Pins the worker-shutdown classification (pkg/runnerfailure parity,
 * issue #776): the two message shapes that mean "infrastructure
 * interrupted the turn", matched exactly as the Go and Java control
 * planes match them, and the honest status.error copy every edition
 * shares. The invoke workflow's recovery loop and failure-path persist
 * both key on this classifier.
 */
import { ApplicationFailure, CancelledFailure } from "@temporalio/common";
import { describe, expect, it } from "vitest";

import {
  isWorkerShutdown,
  WORKER_SHUTDOWN_STATUS_ERROR,
} from "../runner-failure.js";

describe("isWorkerShutdown", () => {
  it("matches the runner's own shutdown classification in a cancelled failure", () => {
    expect(
      isWorkerShutdown(
        new CancelledFailure("Activity cancelled (worker shutdown, not user pause)"),
      ),
    ).toBe(true);
  });

  it("matches loose markers only inside CANCELLED failures (runner-authored text)", () => {
    expect(isWorkerShutdown(new CancelledFailure("shutting down"))).toBe(true);
    // The same loose text in a NON-cancelled failure could be agent/tool
    // output — it must NOT classify.
    expect(isWorkerShutdown(new Error("shutting down"))).toBe(false);
  });

  it("matches the Temporal TS worker's drain text in application failures", () => {
    expect(
      isWorkerShutdown(
        ApplicationFailure.create({
          message:
            "Worker is shutting down and this activity did not complete in time",
        }),
      ),
    ).toBe(true);
  });

  it("matches the drain text anywhere in the cause chain", () => {
    const wrapped = new Error("activity 'ExecuteDeepAgent' failed", {
      cause: ApplicationFailure.create({
        message: "Worker is shutting down and this activity did not complete in time",
      }),
    });
    expect(isWorkerShutdown(wrapped)).toBe(true);
  });

  it("does not classify ordinary failures", () => {
    expect(isWorkerShutdown(new Error("agent blew up"))).toBe(false);
    expect(
      isWorkerShutdown(ApplicationFailure.create({ message: "boom" })),
    ).toBe(false);
    expect(isWorkerShutdown(undefined)).toBe(false);
  });

  it("pins the honest cross-edition status.error copy", () => {
    expect(WORKER_SHUTDOWN_STATUS_ERROR).toBe(
      "Execution interrupted: runner worker was shut down. Retry or resume.",
    );
  });
});
